pids/pids.man
.\" format with nroff|troff|groff -man
.fp 5 CW
.nr mI 0
.de mI
.if \\n(mI>\\$1 \{
.nr mI \\n(mI-1
.RE
.mI \\$1
.\}
.if \\n(mI<\\$1 \{
.nr mI \\n(mI+1
.RS
.mI \\$1
.\}
..
.de H1
.mI 1
.TP
\fB\\$1\fP
..
.de H2
.mI 2
.TP
\fB\\$1\fP
..
.de H3
.mI 3
.TP
\fB\\$1\fP
..
.de H4
.mI 4
.TP
\fB\\$1\fP
..
.de OP
.mI 0
.ie !'\\$1'-' \{
.ds mO \\fB\\-\\$1\\fP
.ds mS ,\\0
.\}
.el \{
.ds mO \\&
.ds mS \\&
.\}
.ie '\\$2'-' \{
.if !'\\$4'-' .as mO \\0\\fI\\$4\\fP
.\}
.el \{
.as mO \\*(mS\\fB\\-\\-\\$2\\fP
.if !'\\$4'-' .as mO =\\fI\\$4\\fP
.\}
.TP
\\*(mO
..
.de FN
.mI 0
.TP
\\$1 \\$2
..
.TH pids 1 "2008-04-01" AST
.SH NAME
pids \- list calling shell process ids
.SH SYNOPSIS
\fBpids\fP\ [\ \fIoptions\fP\ ]
.SH DESCRIPTION
When invoked as a shell builtin, \fBpids\fP lists one or more of the calling
process ids determined by \fBgetpid\fP(2), \fBgetppid\fP(2), \fBgetpgrp\fP(2),
\fBtcgetpgrp\fP(2) and \fBgetsid\fP(2)\&. Unknown or invalid ids have the value
\fB\-1\fP\&.
.SH OPTIONS
.OP f format string format PID=%(pid)d PPID=%(ppid)d PGID=%(pgid)d TID=%(tid)d
SID=%(sid)d
List the ids specified by \fIformat\fP\&. \fIformat\fP follows \fBprintf\fP(3)
conventions, except that \fBsfio\fP(3) inline ids are used instead of
arguments:
%[\-+][\fIwidth\fP[\&.\fIprecis\fP[\&.\fIbase\fP]]](\fIid\fP)\fIchar\fP\&. The
supported \fIid\fPs are:
.H3 pid
The process id\&.
.H3 pgid
The process group id\&.
.H3 ppid
The parent process id\&.
.H3 tid|tty
The controlling terminal id\&.
.H3 sid
The session id\&.
.PP
The default value is \fBPID=%(pid)d PPID=%(ppid)d PGID=%(pgid)d TID=%(tid)d
SID=%(sid)d\fP\&.
.SH SEE\ ALSO
\fBgetpid\fP(2), \fBgetppid\fP(2), \fBgetpgrp\fP(2), \fBtcgetpgrp\fP(2),
\fBgetsid\fP(2)
.SH IMPLEMENTATION
.H1 version
pids (AT&T Research) 2008\-04\-01
.\"_
.\"_


rshe/CHANGES
RSHE



= Dave Morano, 95/08/22
version '0a'

This program was originally written.  It was inspired by the
'bopenrcmd' as part of the "Basic I/O" (BIO) package.


= Dave Morano, 95/10/12
version '0b'

1) The meaning of the 'x' program invocation option was changed
   to match other such modern "remote execution" programs such
   as 'rex'.





pids/pids.c
/***********************************************************************
*                                                                      *
*               This software is part of the ast package               *
*          Copyright (c) 1992-2008 AT&T Intellectual Property          *
*                      and is licensed under the                       *
*                  Common Public License, Version 1.0                  *
*                    by AT&T Intellectual Property                     *
*                                                                      *
*                A copy of the License is available at                 *
*            http://www.opensource.org/licenses/cpl1.0.txt             *
*         (with md5 checksum 059e8cd6165cb4c31e351f2b69388fd9)         *
*                                                                      *
*              Information and Software Systems Research               *
*                            AT&T Research                             *
*                           Florham Park NJ                            *
*                                                                      *
*                 Glenn Fowler <gsf@research.att.com>                  *
*                  David Korn <dgk@research.att.com>                   *
*                                                                      *
***********************************************************************/
#pragma prototyped

#define FORMAT		"PID=%(pid)d PPID=%(ppid)d PGID=%(pgid)d TID=%(tid)d SID=%(sid)d"

static const char usage[] =
"[-?\n@(#)$Id: pids (AT&T Research) 2008-04-01 $\n]"
/* USAGE_LICENSE */
"[+NAME?pids - list calling shell process ids]"
"[+DESCRIPTION?When invoked as a shell builtin, \bpids\b lists one or "
    "more of the calling process ids determined by \bgetpid\b(2), "
    "\bgetppid\b(2), \bgetpgrp\b(2), \btcgetpgrp\b(2) and \bgetsid\b(2). "
    "Unknown or invalid ids have the value \b-1\b.]"
"[f:format?List the ids specified by \aformat\a. \aformat\a follows "
    "\bprintf\b(3) conventions, except that \bsfio\b(3) inline ids are used "
    "instead of arguments: "
    "%[-+]][\awidth\a[.\aprecis\a[.\abase\a]]]]]](\aid\a)\achar\a. The "
    "supported \aid\as are:]:[format:=" FORMAT "]"
    "{"
        "[+pid?The process id.]"
        "[+pgid?The process group id.]"
        "[+ppid?The parent process id.]"
        "[+tid|tty?The controlling terminal id.]"
        "[+sid?The session id.]"
    "}"
"[+SEE ALSO?\bgetpid\b(2), \bgetppid\b(2), \bgetpgrp\b(2), "
    "\btcgetpgrp\b(2), \bgetsid\b(2)]"
;



/* local defines */

#ifndef	ERROR_CATALOG
#define	ERROR_CATALOG	"libast"
#endif





#include <cmd.h>
#include <ast_tty.h>
#include <sfdisc.h>

/*
 * sfkeyprintf() lookup
 * handle==0 for heading
 */

static int
key(void* handle, Sffmt_t* fp, const char* arg, char** ps, Sflong_t* pn)
{
	register char*	s;
	int		fd;
	long		tid;

	if (!(s = fp->t_str) || streq(s, "pid"))
		*pn = getpid();
	else if (streq(s, "pgid"))
		*pn = getpgid(0);
	else if (streq(s, "ppid"))
		*pn = getppid();
	else if (streq(s, "tid") || streq(s, "tty"))
	{
		for (fd = 0; fd < 3; fd++)
			if ((tid = tcgetpgrp(fd)) >= 0)
				break;
		*pn = tid;
	}
	else if (streq(s, "sid"))
		*pn = getsid(0);
	else if (streq(s, "format"))
		*ps = (char*)handle;
	else
	{
		error(2, "%s: unknown format identifier", s);
		return 0;
	}
	return 1;
}

int
b_pids(int argc, char** argv, void* context)
{
	char*			format = 0;

	cmdinit(argc, argv, context, ERROR_CATALOG, 0);
	for (;;)
	{
		switch (optget(argv, usage))
		{
		case 'f':
			format = opt_info.arg;
			continue;
		case '?':
			error(ERROR_USAGE|4, "%s", opt_info.arg);
			continue;
		case ':':
			error(2, "%s", opt_info.arg);
			continue;
		}
		break;
	}
	argv += opt_info.index;
	if (error_info.errors || *argv)
		error(ERROR_USAGE|4, "%s", optusage(NiL));
	if (!format)
		format = FORMAT;
	sfkeyprintf(sfstdout, format, format, key, NiL);
	sfprintf(sfstdout, "\n");
	return 0;
}
